File: java-backend/src/main/java/com/study/ws/WebSocketAuthInterceptor.java
package com.study.ws;

import com.study.security.JwtService;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import org.springframework.http.server.ServerHttpRequest;
import org.springframework.http.server.ServerHttpResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketHandler;
import org.springframework.web.socket.server.HandshakeInterceptor;

/** WS 握手鉴权:从 ?token= 取 JWT 校验(浏览器 WS 不能带 header),通过则把 username 放进会话属性。 */
@Component
public class WebSocketAuthInterceptor implements HandshakeInterceptor {

    private final JwtService jwtService;

    public WebSocketAuthInterceptor(JwtService jwtService) {
        this.jwtService = jwtService;
    }

    @Override
    public boolean beforeHandshake(ServerHttpRequest request, ServerHttpResponse response,
                                   WebSocketHandler wsHandler, Map<String, Object> attributes) {
        String token = tokenFrom(request.getURI().getQuery());
        if (token == null) {
            return false;
        }
        try {
            attributes.put("username", jwtService.parse(token).get("username", String.class));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    @Override
    public void afterHandshake(ServerHttpRequest request, ServerHttpResponse response,
                              WebSocketHandler wsHandler, Exception exception) {
        // 无需处理
    }

    private String tokenFrom(String query) {
        if (query == null) {
            return null;
        }
        for (String part : query.split("&")) {
            if (part.startsWith("token=")) {
                return URLDecoder.decode(part.substring("token=".length()), StandardCharsets.UTF_8);
            }
        }
        return null;
    }
}

File: kotlin-backend/src/main/kotlin/com/study/ws/WebSocketAuthInterceptor.kt
package com.study.ws

import com.study.security.JwtService
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import org.springframework.http.server.ServerHttpRequest
import org.springframework.http.server.ServerHttpResponse
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketHandler
import org.springframework.web.socket.server.HandshakeInterceptor

/** WS 握手鉴权:从 ?token= 取 JWT 校验(浏览器 WS 不能带 header),通过则把 username 放进会话属性。 */
@Component
class WebSocketAuthInterceptor(private val jwtService: JwtService) : HandshakeInterceptor {

    override fun beforeHandshake(
        request: ServerHttpRequest,
        response: ServerHttpResponse,
        wsHandler: WebSocketHandler,
        attributes: MutableMap<String, Any>,
    ): Boolean {
        val token = tokenFrom(request.uri.query) ?: return false
        return try {
            attributes["username"] = jwtService.parse(token).get("username", String::class.java) ?: "匿名"
            true
        } catch (e: Exception) {
            false
        }
    }

    override fun afterHandshake(
        request: ServerHttpRequest,
        response: ServerHttpResponse,
        wsHandler: WebSocketHandler,
        exception: Exception?,
    ) {
        // 无需处理
    }

    private fun tokenFrom(query: String?): String? =
        query?.split("&")
            ?.firstOrNull { it.startsWith("token=") }
            ?.let { URLDecoder.decode(it.substring("token=".length), StandardCharsets.UTF_8) }
}
